src/ossl/ecc.rs

```diff
@@ -709,3 +716,149 @@ impl Verify for EccOperation {
         Ok(self.output_len)
     }
 }
+
+pub const NONE_NAME: &[u8; 1] = b"\0";
+fn ckd_to_type_name(mech: CK_EC_KDF_TYPE) -> *const c_char {
+    (match mech {
+        CKD_SHA1_KDF | CKD_SHA224_KDF | CKD_SHA256_KDF | CKD_SHA384_KDF
+        | CKD_SHA512_KDF | CKD_SHA3_224_KDF | CKD_SHA3_256_KDF
+        | CKD_SHA3_384_KDF | CKD_SHA3_512_KDF => OSSL_KDF_NAME_X963KDF.as_ptr(),
+        _ => NONE_NAME.as_ptr(),
+    }) as *const c_char
+}
+
+fn new_pkey_ctx_from_obj(pkey: &mut EvpPkey) -> KResult<EvpPkeyCtx> {
+    Ok(EvpPkeyCtx::from_ptr(unsafe {
+        EVP_PKEY_CTX_new_from_pkey(
+            get_libctx(),
+            pkey.as_mut_ptr(),
+            std::ptr::null(),
+        )
+    })?)
+}
+
+impl Derive for ECDHOperation<'_> {
+    fn derive(
+        &mut self,
+        key: &Object,
+        template: &[CK_ATTRIBUTE],
+        _mechanisms: &Mechanisms,
+        objfactories: &ObjectFactories,
+    ) -> KResult<(Object, usize)> {
+        if self.finalized {
+            return err_rv!(CKR_OPERATION_NOT_INITIALIZED);
+        }
+        self.finalized = true;
+
+        let mut params = OsslParam::with_capacity(6).set_zeroize().add_int(
+            name_as_char(OSSL_EXCHANGE_PARAM_EC_ECDH_COFACTOR_MODE),
+            if self.mech == CKM_ECDH1_COFACTOR_DERIVE {
+                1
+            } else {
+                -1
+            } as i32,
+        )?;
+        let factory =
+            objfactories.get_obj_factory_from_key_template(template)?;
+
+        let keylen = match template.iter().find(|x| x.type_ == CKA_VALUE_LEN) {
+            Some(a) => a.to_ulong()? as usize,
+            None => {
```

Comment:
I think you may want to add a test where you do not specify CKA_VALUE_LEN so we ensure this code path works as intended.

Reply:
Will do.

Reply:
This looks like this part was wrong. We should not look at hash here, but the plain ECDH max is the curve bits and X9.63 requires the length explicitly so there is no default max length ...